File: go/test/alternate_list_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/tvmaze-sdk/go"
	"github.com/voxgig-sdk/tvmaze-sdk/go/core"

	vs "github.com/voxgig-sdk/tvmaze-sdk/go/utility/struct"
)

func TestAlternateListEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.AlternateList(nil)
		if ent == nil {
			t.Fatal("expected non-nil AlternateListEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := alternate_listBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"list", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "alternate_list." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set TVMAZE_TEST_ALTERNATE_LIST_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		alternateListRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.alternate_list", setup.data)))
		var alternateListRef01Data map[string]any
		if len(alternateListRef01DataRaw) > 0 {
			alternateListRef01Data = core.ToMapAny(alternateListRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = alternateListRef01Data

		// LIST
		alternateListRef01Ent := client.AlternateList(nil)
		alternateListRef01Match := map[string]any{
			"show_id": setup.idmap["show01"],
		}

		alternateListRef01ListResult, err := alternateListRef01Ent.List(alternateListRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, alternateListRef01ListOk := alternateListRef01ListResult.([]any)
		if !alternateListRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", alternateListRef01ListResult)
		}

		// LOAD
		alternateListRef01MatchDt0 := map[string]any{
			"id": alternateListRef01Data["id"],
		}
		alternateListRef01DataDt0Loaded, err := alternateListRef01Ent.Load(alternateListRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		alternateListRef01DataDt0LoadResult := core.ToMapAny(alternateListRef01DataDt0Loaded)
		if alternateListRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if alternateListRef01DataDt0LoadResult["id"] != alternateListRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func alternate_listBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "alternate_list", "AlternateListTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read alternate_list test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse alternate_list test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"alternate_list01", "alternate_list02", "alternate_list03", "show01", "show02", "show03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("TVMAZE_TEST_ALTERNATE_LIST_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"TVMAZE_TEST_ALTERNATE_LIST_ENTID": idmap,
		"TVMAZE_TEST_LIVE":      "FALSE",
		"TVMAZE_TEST_EXPLAIN":   "FALSE",
		"TVMAZE_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["TVMAZE_TEST_ALTERNATE_LIST_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["TVMAZE_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["TVMAZE_APIKEY"],
			},
			extra,
		})
		client = sdk.NewTvmazeSDK(core.ToMapAny(mergedOpts))
	}

	live := env["TVMAZE_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["TVMAZE_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: ts/test/entity/alternate_list/AlternateListEntity.test.ts

const envlocal = __dirname + '/../../../.env.local'
require('dotenv').config({ quiet: true, path: [envlocal] })

import Path from 'node:path'
import * as Fs from 'node:fs'

import { test, describe, afterEach } from 'node:test'
import assert from 'node:assert'


import { TvmazeSDK, BaseFeature, stdutil } from '../../..'

import {
  envOverride,
  liveDelay,
  makeCtrl,
  makeMatch,
  makeReqdata,
  makeStepData,
  makeValid,
  maybeSkipControl,
} from '../../utility'


describe('AlternateListEntity', async () => {

  // Per-test live pacing. Delay is read from sdk-test-control.json's
  // `test.live.delayMs`; only sleeps when TVMAZE_TEST_LIVE=TRUE.
  afterEach(liveDelay('TVMAZE_TEST_LIVE'))

  test('instance', async () => {
    const testsdk = TvmazeSDK.test()
    const ent = testsdk.AlternateList()
    assert(null != ent)
  })


  test('basic', async (t) => {

    const live = 'TRUE' === process.env.TVMAZE_TEST_LIVE
    for (const op of ['list', 'load']) {
      if (maybeSkipControl(t, 'entityOp', 'alternate_list.' + op, live)) return
    }

    const setup = basicSetup()
    // The basic flow consumes synthetic IDs and field values from the
    // fixture (entity TestData.json). Those don't exist on the live API.
    // Skip live runs unless the user provided a real ENTID env override.
    if (setup.syntheticOnly) {
      t.skip('live entity test uses synthetic IDs from fixture — set TVMAZE_TEST_ALTERNATE_LIST_ENTID JSON to run live')
      return
    }
    const client = setup.client
    const struct = setup.struct

    const isempty = struct.isempty
    const select = struct.select

    let alternate_list_ref01_data = Object.values(setup.data.existing.alternate_list)[0] as any

    // LIST
    const alternate_list_ref01_ent = client.AlternateList()
    const alternate_list_ref01_match: any = {}
    alternate_list_ref01_match['show_id'] = setup.idmap['show01']

    const alternate_list_ref01_list = await alternate_list_ref01_ent.list(alternate_list_ref01_match)


    // LOAD
    const alternate_list_ref01_match_dt0: any = {}
    alternate_list_ref01_match_dt0.id = alternate_list_ref01_data.id
    const alternate_list_ref01_data_dt0 = await alternate_list_ref01_ent.load(alternate_list_ref01_match_dt0)
    assert(alternate_list_ref01_data_dt0.id === alternate_list_ref01_data.id)


  })
})



function basicSetup(extra?: any) {
  // TODO: fix test def options
  const options: any = {} // null

  // TODO: needs test utility to resolve path
  const entityDataFile =
    Path.resolve(__dirname, 
      '../../../../.sdk/test/entity/alternate_list/AlternateListTestData.json')

  // TODO: file ready util needed?
  const entityDataSource = Fs.readFileSync(entityDataFile).toString('utf8')

  // TODO: need a xlang JSON parse utility in voxgig/struct with better error msgs
  const entityData = JSON.parse(entityDataSource)

  options.entity = entityData.existing

  let client = TvmazeSDK.test(options, extra)
  const struct = client.utility().struct
  const merge = struct.merge
  const transform = struct.transform

  let idmap = transform(
    ['alternate_list01','alternate_list02','alternate_list03','show01','show02','show03'],
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  // Detect whether the user provided a real ENTID JSON via env var. The
  // basic flow consumes synthetic IDs from the fixture file; without an
  // override those synthetic IDs reach the live API and 4xx. Surface this
  // to the test so it can skip rather than fail.
  const idmapEnvVal = process.env['TVMAZE_TEST_ALTERNATE_LIST_ENTID']
  const idmapOverridden = null != idmapEnvVal && idmapEnvVal.trim().startsWith('{')

  const env = envOverride({
    'TVMAZE_TEST_ALTERNATE_LIST_ENTID': idmap,
    'TVMAZE_TEST_LIVE': 'FALSE',
    'TVMAZE_TEST_EXPLAIN': 'FALSE',
  })

  idmap = env['TVMAZE_TEST_ALTERNATE_LIST_ENTID']

  const live = 'TRUE' === env.TVMAZE_TEST_LIVE

  if (live) {
    client = new TvmazeSDK(merge([
      {
      },
      extra
    ]))
  }

  const setup = {
    idmap,
    env,
    options,
    client,
    struct,
    data: entityData,
    explain: 'TRUE' === env.TVMAZE_TEST_EXPLAIN,
    live,
    syntheticOnly: live && !idmapOverridden,
    now: Date.now(),
  }

  return setup
}
  
